planner/cascades/transformation_rules.go

```diff
@@ -1952,3 +1954,72 @@ func (r *TransformAggregateCaseToSelection) isTwoOrThreeArgCase(expr expression.
 	}
 	return scalarFunc.FuncName.L == ast.Case && (len(scalarFunc.GetArgs()) == 2 || len(scalarFunc.GetArgs()) == 3)
 }
+
+// EliminateTableDualBelowUnionAll will eliminate LogicalTableDual below LogicalUnionAll.
+// This rule might be useful for pruning partitions of partitioned table.
+type EliminateTableDualBelowUnionAll struct {
+	baseRule
+}
+
+// NewRuleEliminateTableDualBelowUnionAll creates a new Transformation EliminateTableDualBelowUnionAll.
+func NewRuleEliminateTableDualBelowUnionAll() Transformation {
+	rule := &EliminateTableDualBelowUnionAll{}
+	rule.pattern = memo.NewPattern(memo.OperandUnionAll, memo.EngineTiDBOnly)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to remove TableDual below UnionAll.
+func (r *EliminateTableDualBelowUnionAll) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	expr := old.GetExpr()
+	newChildren := make([]*memo.Group, 0, len(expr.Children))
+	for _, childGroup := range expr.Children {
+		if !r.groupContainsTableDual(childGroup) {
+			newChildren = append(newChildren, childGroup)
+		}
+	}
+	if len(newChildren) == len(expr.Children) {
+		// There is no TableDual below UnionAll.
+		return nil, false, false, nil
+	}
```

Comment:
Can we move these to function `Match`?

Reply:
The `newChildren` will be used again. If we move this to `Match`, we still have to collect the `newChildren` here.

Actually, the method `Match` is a redundant design now. We don't even need this method in the interface of `Transformation`.